addons/udes_advanced_packaging/models/stock_picking.py

```diff
@@ -0,0 +1,156 @@
+# -*- coding: utf-8 -*-
+from collections import defaultdict
+
+from odoo import models, _
+from odoo.exceptions import ValidationError
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def _should_reserve_full_packages(self):
+        """Method to determine if picking should reserve entire packages"""
+        self.ensure_one()
+        return self.picking_type_id.u_reserve_as_packages
+
+    def _reserve_full_packages(self):
+        """
+        If the picking type of the picking in self has full package
+        reservation enabled, whole packages will be reserved.
+        """
+        Quant = self.env["stock.quant"]
+        MoveLine = self.env["stock.move.line"]
+
+        # do not reserve full packages when bypass_reserve_full packages
+        # is set in the context as True
+        if not self.env.context.get("bypass_reserve_full_packages"):
+            for picking in self.filtered(lambda p: p._should_reserve_full_packages()):
+                all_quants = Quant.browse()
+                remaining_qtys = defaultdict(int)
+
+                # get all packages
+                packages = picking.mapped("move_line_ids.package_id")
+                for package in packages:
+                    move_lines = picking.mapped("move_line_ids").filtered(
+                        lambda ml: ml.package_id == package
+                    )
+                    pack_products = frozenset(package._get_all_products_quantities().items())
+                    mls_products = frozenset(move_lines._get_all_products_quantities().items())
+                    if pack_products != mls_products:
+                        # move_lines do not match the quants
+                        pack_mls = MoveLine.search(
+                            [
+                                ("package_id", "child_of", package.id),
+                                ("state", "not in", ["done", "cancel"]),
+                            ]
+                        )
+                        other_pickings = pack_mls.mapped("picking_id") - picking
```

Comment:
Like boss pete said, the use of mapped everywhere is no longer needed in all places.

Reply:
Fixed in latest commit